vsintegration/src/FSharp.Editor/AutomaticCompletion/BraceCompletionSessionProvider.fs

```diff
@@ -0,0 +1,410 @@
+                            editorOperations.AddAfterTextBufferChangePrimitive()
+                            undo.Complete()
+                
+        member __.PostOverType() = ()
+
+        member this.PreTab handledCommand =
+            handledCommand <- false
+
+            if not this.HasForwardTyping then
+                handledCommand <- true
+
+                use undo = this.CreateUndoTransaction()
+
+                editorOperations.AddBeforeTextBufferChangePrimitive()
+                this.MoveCaretToClosingPoint()
+                editorOperations.AddAfterTextBufferChangePrimitive()
+                undo.Complete()
+
+        member __.PreReturn handledCommand =
+            handledCommand <- false
+
+        member this.PostReturn() =
+            let caretPos = getCaretPosition this
+
+            if caretPos.HasValue then
+                let closingSnapshotPoint = closingPoint.GetPoint(subjectBuffer.CurrentSnapshot)
+
+                if closingSnapshotPoint.Position > 0 && this.HasNoForwardTyping(caretPos.Value, closingSnapshotPoint.Subtract(1)) then
+                    session.AfterReturn(this, CancellationToken.None)
+                
+        member __.Finish() = ()
+
+        member __.PostTab() = ()
+
+        member __.PreDelete handledCommand =
+            handledCommand <- false
+        
+        member __.PostDelete() = ()
+
+        member __.OpeningBrace = openingBrace
+
+        member __.ClosingBrace = closingBrace
+
+        member __.OpeningPoint = openingPoint
+
+        member __.ClosingPoint = closingPoint
+
+        member __.SubjectBuffer = subjectBuffer
+
+        member __.TextView = textView
+
+module private Parenthesis =
+
+    [<Literal>]
+    let OpenCharacter = '('
+
+    [<Literal>]
+    let CloseCharacter = ')'
+
+type internal ParenthesisCompletionSession () =
+    
+    interface IEditorBraceCompletionSession with
+
+        member this.AfterReturn(_session, _cancellationToken) = 
+            ()
+
+        member this.AfterStart(_session, _cancellationToken) = 
+            ()
+
+        member this.AllowOverType(_session, _cancellationToken) = 
+            // TODO: Implement this for F#
+            true
+
+        member this.CheckOpeningPoint(_session, _cancellationToken) = 
+            // TODO: Implement this for F#
+            true 
+
+[<ExportLanguageService(typeof<IEditorBraceCompletionSessionFactory>, FSharpConstants.FSharpLanguageName)>]
+type internal FSharpEditorBraceCompletionSessionFactory () =
+    inherit ForegroundThreadAffinitizedObject ()
+
+    member __.IsSupportedOpeningBrace openingBrace =
+        match openingBrace with
+        | Parenthesis.OpenCharacter -> true
+        | _ -> false
+
+    member this.CheckCodeContext(_document, _position, _openingBrace, _cancellationToken) =
+        this.AssertIsForeground();
+
+        // TODO: We need to know if we are inside a F# comment. If we are, then don't do automatic completion.
```

Comment:
This should also apply when in the middle of a string literal.

Reply:
Indeed. Adding that to the list.